kernel/thread.c

```diff
@@ -348,40 +410,172 @@ long myst_syscall_wait4(
+        for (p = _zombies_head; p; p = p->znext)
         {
-            bool match = false;
-
-            if (!myst_is_process_thread(p))
-                continue;
-
-            if (pid > 0) /* wait for a specific child process */
-            {
-                match = p->pid == pid;
-            }
-            else if (pid == -1) /* wait for any child process */
-            {
-                match = true;
-            }
+            myst_assume(myst_is_process_thread(p));
 
-            if (match)
+            if (_wait_matcher("zombie", pid, process, p))
             {
                 if (wstatus)
-                    *wstatus = (p->exit_status << 8);
-
+                {
+                    *wstatus = 0;
+
+                    if (p->terminating_signum)
+                    {
+                        *wstatus = (p->terminating_signum & 0x7f);
+                    }
+                    else
+                    {
+                        *wstatus = ((p->exit_status & 0xff) << 8);
+                    }
+#ifdef TRACE
+                    printf("*** WSTATUS DETAILS: %d\n", *wstatus);
+                    printf(
+                        "    WIFEXITED (exited through exit(), _exit() or "
+                        "return from main): %s\n",
+                        WIFEXITED(*wstatus) ? "true" : "false");
+                    printf(
+                        "    WEXITSTATUS (return code via exit(), _exit() or "
+                        "return from main): %u\n",
+                        WEXITSTATUS(*wstatus));
+                    printf(
+                        "    WIFSIGNALED (exited via signal): %s\n",
+                        WIFSIGNALED(*wstatus) ? "true" : "false");
+                    printf(
+                        "    WTERMSIG (signal code to terminate): %u\n",
+                        WTERMSIG(*wstatus));
+                    printf(
+                        "    WCOREDUMP (signaled and dumped core, not "
+                        "supported): %s\n",
+                        WCOREDUMP(*wstatus) ? "true" : "false");
+                    printf(
+                        "    WIFSTOPPED (signaled to stop, not supported): "
+                        "%s\n",
+                        WIFSTOPPED(*wstatus) ? "true" : "false");
+                    if (WIFSTOPPED(*wstatus))
+                        printf(
+                            "    WSTOPSIG (signal code to stop, not "
+                            "supported): "
+                            "%u\n",
+                            WSTOPSIG(*wstatus));
+                    printf(
+                        "    WIFCONTINUED (signaled to resumed after stop, not "
+                        "supported): %s\n",
+                        WIFCONTINUED(*wstatus) ? "true" : "false");
+#endif
+                }
                 ret = p->pid;
+
+                // remove from zombie list
+                if (p->zprev == NULL)
+                {
+                    _zombies_head = p->znext;
```

Comment:
Consider using "myst_list_t" to manage zombies. We can replace line 470-484 with a simple `myst_list_remove`.

Reply:
we can consider that for a future update.